### PR TITLE
Created set_snark_work commands

### PR DESCRIPTION
**Pull request by wu-s-john**
Here are the effects of setting a snark worker on the cli

# CLI

## Before
Got status before running `set_snark_worker` command

```
Coda daemon status
-----------------------------------

Max observed block length:  1
Local uptime:               9s
Git SHA-1:                  7c750b132edc8099c01b1b3b7e932a0a190f5bf1
Configuration directory:    /Users/johnwu/.coda-config
Peers:                      0
User_commands sent:         0
SNARK worker running:       false
Sync status:                Bootstrap
Block producers running:    0
Consensus time now:         epoch=14644, slot=526
Consensus mechanism:        proof_of_stake
Consensus configuration:    
	Delta:                     3
	k:                         24
	c:                         8
	c * k:                     192
	Slots per epoch:           576
	Slot duration:             2s
	Epoch duration:            19m12s
	Acceptable network delay:  6s
```

## After 

Ran this command
```
coda.exe client  set-snark-worker -address 8QnLYnzAkhx3moXHHBBpK418LGCEiQZzYDgKggntsNkZqsUJLFos474Mzoy5mWJP9S
```
Then got the status
```
Coda daemon status
-----------------------------------

Max observed block length:  1
Local uptime:               3m14s
Git SHA-1:                  7c750b132edc8099c01b1b3b7e932a0a190f5bf1
Configuration directory:    /Users/johnwu/.coda-config
Peers:                      0
User_commands sent:         0
SNARK worker running:       true
Sync status:                Bootstrap
Block producers running:    0
Consensus time now:         epoch=14644, slot=507
Consensus mechanism:        proof_of_stake
Consensus configuration:    
	Delta:                     3
	k:                         24
	c:                         8
	c * k:                     192
	Slots per epoch:           576
	Slot duration:             2s
	Epoch duration:            19m12s
	Acceptable network delay:  6s
```

Here are the effects of setting a snark worker on graphql

# Graphql

Ran this command:

```
mutation {setSnarkWorker (input : {wallet: "8QnLVikbt8DRYhUuN7m9DpAb2jGGpH14vXdTUqGsFZNENwiPnwp9Hts1m6Qx7C4tAr"}) {lastSnarkWorker }
}
```

Got the following output:

```
{
  "data": {
    "setSnarkWorker": {
      "lastSnarkWorker": "8QnLXcu2tXnd9WJ3fuwaC52jUpMjCi4kV8c1CeVsJ5dqDwZNN96gbYAdHwhoS8fhmt"
    }
  }
}
```

@Schmavery  sorry for not using graphql-ppx. Had a difficult time trying to compile a mutation command. Not sure how to troubleshoot it.

Checklist:

- [X] Tests were added for the new behavior
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them:

Closes #0000
Closes #0000

### CODE REVIEW

File: src/app/cli/src/graphql.ml [hunk 1]
```diff
@@ -689,6 +698,14 @@ module Types = struct
                 "Public keys of wallets you wish to stake - these must be \
                  wallets that are in ownedWallets" ]
 
+    let set_snark_worker =
+      obj "SetSnarkWorkerInput" ~coerce:Fn.id
+        ~fields:
+          [ arg "wallet" ~typ:public_key_arg
+              ~doc:
+                "A public key you wish to get reward for doing snark work - \
```
Comment (bkase) [inline src/app/cli/src/graphql.ml]:
```
Public key you wish to start snark-working on; null to stop doing any snark work
```

File: src/app/cli/src/graphql.ml [hunk 2]
```diff
@@ -1253,13 +1270,24 @@ module Mutations = struct
         ignore @@ Coda_commands.replace_proposers coda pks ;
         Public_key.Compressed.Set.to_list old_propose_keys )
 
+  let set_snark_worker =
+    field "setSnarkWorker"
+      ~doc:"Set key you wish to get rewarded for doing snark work"
```
Comment (bkase) [inline src/app/cli/src/graphql.ml]: `Set key you wish to snark work with or disable snark working`

**Comment by bkase** [comment]
It's important to update the graphql schema before landing this anyway otherwise the docs won't be updated either.

**Comment by bkase** [comment]
Oh this needs to land before it will work! https://github.com/CodaProtocol/coda/pull/3175

**Comment by jkrauska** [comment]
@wu-s-john  can you match the syntax of the block proposer and LIST the key that you're using for the snark-worker in status?

**Comment by wu-s-john** [comment]
@jkrauska 

Done:

Here is the output now:


```
Coda daemon status
-----------------------------------

Max observed block length:  1
Local uptime:               20s
Git SHA-1:                  7c750b132edc8099c01b1b3b7e932a0a190f5bf1
Configuration directory:    /Users/johnwu/.coda-config
Peers:                      0
User_commands sent:         0
SNARK worker:               None
Sync status:                Bootstrap
Block producers running:    0
Consensus time now:         epoch=14692, slot=29
Consensus mechanism:        proof_of_stake
Consensus configuration:    
	Delta:                     3
	k:                         24
	c:                         8
	c * k:                     192
	Slots per epoch:           576
	Slot duration:             2s
	Epoch duration:            19m12s
	Acceptable network delay:  6s
```

When the I set the snark worker, here is the output:

```
src/_build/default/app/cli/src/coda.exe client set-snark-worker -address "8QnLVb6L8r4WVJzFZXGcAr1zTqpAo4u4bXU1G4ThXgnndRJF3wUE1jJdVELtTcC5x7 
```

```
Coda daemon status
-----------------------------------

Max observed block length:  1
Local uptime:               4m4s
Git SHA-1:                  7c750b132edc8099c01b1b3b7e932a0a190f5bf1
Configuration directory:    /Users/johnwu/.coda-config
Peers:                      0
User_commands sent:         0
SNARK worker:               8QnLVb6L8r4WVJzFZXGcAr1zTqpAo4u4bXU1G4ThXgnndRJF3wUE1jJdVELtTcC5x7
Sync status:                Bootstrap
Block producers running:    0
Consensus time now:         epoch=14692, slot=141
Consensus mechanism:        proof_of_stake
Consensus configuration:    
	Delta:                     3
	k:                         24
	c:                         8
	c * k:                     192
	Slots per epoch:           576
	Slot duration:             2s
	Epoch duration:            19m12s
	Acceptable network delay:  6s

```

**Comment by cmr** [comment]
This is a good first step. We will want to have multiple SNARK workers at some point. This GraphQL API only allows for one at a time. I don't care about the implementation supporting multiple workers yet, but I think we should upgrade some scalars into lists to support that in the future.

**Comment by wu-s-john** [comment]
> The code looks fine. At some point (soon probably) we are going to want multiple snark workers. I don't care about the implementation supporting multiple workers right now, but I think we should promote some of the scalars to lists so the API can support it in the future.

I don't think we should this. We do not have support for this yet. Once we add the support to have multiple snark workers, we can then change the API accordingly. It would not be hard to do.

**Pull request merged**